enhancements/installer/storage-class-encrypted.md

```diff
@@ -0,0 +1,308 @@
+
+AWS, Azure & GCP will be added as new valid values for the `CSIDriverType` union discriminator
+in the CSIDriverConfigSpec:
+
+```go
+// CSIDriverType indicates type of CSI driver being configured.
+// +kubebuilder:validation:Enum="";vSphere
+type CSIDriverType string
+
+const (
+  AWSDriverType     CSIDriverType = "aws"
+  AzureDriverType   CSIDriverType = "azure"
+  GCPDriverType     CSIDriverType = "gcp"
+  VSphereDriverType CSIDriverType = "vSphere" // existing platform
+)
+
+// CSIDriverConfigSpec defines configuration spec that can be
+// used to optionally configure a specific CSI Driver.
+// +union
+type CSIDriverConfigSpec struct {
+  // driverType indicates type of CSI driver for which the
+  // driverConfig is being applied to.
+  //
+  // Valid values are:
+  //
+  // * aws
+  // * azure
+  // * gcp
+  // * vSphere
+  //
+  // ---
+  // Consumers should treat unknown values as a NO-OP.
+  //
+  // +kubebuilder:validation:Required
+  // +unionDiscriminator
+  DriverType CSIDriverType `json:"driverType"`
+
+  // AWS is used to configure the AWS CSI driver.
+  // +optional
+  AWS *AWSCSIDriverConfigSpec `json:"aws,omitempty"`
+
+  // Azure is used to configure the Azure CSI driver.
+  // +optional
+  Azure *AzureCSIDriverConfigSpec `json:"azure,omitempty"`
+
+  // GCP is used to configure the GCP CSI driver.
+  // +optional
+  GCP *GCPCSIDriverConfigSpec `json:"gcp,omitempty"`
+
+  ... exisiting vsphere field ... 
+}
+```
+
+ and fields for the encryption keys will be added
+to platform-specific config specs:
+
+#### AWS
+
+```go
+// AWSCSIDriverConfigSpec defines properties that can be configured for the AWS CSI driver.
+type AWSCSIDriverConfigSpec struct {
+    // KMSKeyARN sets the cluster default storage class to encrypt volumes with a user-defined KMS key,
+    // rather than the default KMS key used by AWS.
+    // According to https://docs.aws.amazon.com/service-authorization/latest/reference/list_awskeymanagementservice.html
+    // the format for a KMS key ARN is:
+    // arn:${Partition}:kms:${Region}:${Account}:key/${KeyId}
+    // +optional
+    KMSKeyARN string `json:"kmsKeyARN,omitempty"`
+}
+```
+#### Azure
+
+```go
+// AzureDiskEncryptionSet defines the configuration for a disk encryption set.
+type AzureDiskEncryptionSet struct {
+  // SubscriptionID defines the Azure subscription that contains the disk encryption set.
+  // When omitted, the subscription from the authenticating credentials of the CSI driver
+  // will be used.
+  // +optional 
+  SubscriptionID string `json:"subscriptionId"`
+  
+  // ResourceGroup defines the Azure resource group that contains the disk encryption set.
+  // When omitted, the cluster resource group from the cluster infrastructure object
+  // will be used.
+  // +optional
+  ResourceGroup string `json:"resourceGroup"`
```

Comment:
Fields like this I have marked as optional and defined defaults which the csi driver should supply. It would be simple for the installer to always supply these defaults, so we could make these fields required if there is value in making this API more explicit. 

Users can edit these fields so I find this approach more user friendly. Let me know what you think.

Reply:
IMO, it would be helpful if the installer always supplied these fields whenever `AzureDiskEncryptionSet` is provided. The operator already has a lister for ClusterCSIDriver, but it would take a bit more plumbing to get the ResourceGroup from the Infrastructure object (CSO watches Infra, but azure-disk-csi-driver-operator does not).

Reply:
I am going to update this to make the fields required. I think we can always update to make the fields optional by supplying a sane default, but we can't go from optional -> required: so better to start with required.